poetry.lock

```diff
@@ -1207,22 +1267,17 @@ telegram = ["requests"]
 
 [[package]]
 name = "treq"
-version = "22.2.0"
-description = "High-level Twisted HTTP Client API"
+version = "15.1.0"
```

Comment:
Ugh, poetry is downgrading the `treq` dependency to what's allowed in `pyproject.toml` (`treq = ">=15.1"`) probably because it's the one version that doesn't specify/require a certain `Twisted` version.

We need the newer `treq` version in order to avoid errors like the following:

https://github.com/matrix-org/synapse/runs/7645395562?check_suite_focus=true#step:7:6727
```
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.10/lib/python3.10/site-packages/treq/client.py", line 11, in <module>
    from twisted.python.compat import _PY3, unicode
builtins.ImportError: cannot import name '_PY3' from 'twisted.python.compat' (unknown location)
```

But we can't use the git `trunk` version of `Twisted` (`22.4.0.post0`) because it doesn't satisfy `treqs` requirements in the new versions. It doesn't like the postfix in Twisted version.

```
and treq (21.5.0) depends on Twisted (>=18.7.0), treq (21.5.0) requires twisted (18.7.0 || 18.9.0 || 19.2.0 || 19.2.1 || 19.7.0 || 19.10.0 || 20.3.0 || 21.2.0 || 21.7.0 || 22.1.0 || 22.2.0 || 22.4.0).
  So, because matrix-synapse depends on both twisted (rev trunk) and treq (21.5.0), version solving failed.
```

And we can't override or ignore sub-dependency requirements in Poetry, https://github.com/python-poetry/poetry/issues/697

And Poetry's `allow-prereleases` option doesn't work for non-direct dependencies, https://github.com/python-poetry/poetry/issues/4405

Reply:
Odd and unfortunate.

I wonder if we see this in the "twisted trunk" CI job, https://github.com/matrix-org/synapse/blob/e9ce4d089bbb013f870bbc8d58ec796e8f315eb4/.github/workflows/twisted_trunk.yml

I remember raising an issue related to twisted trunk in poetry for that CI job: https://github.com/python-poetry/poetry/issues/5311 . Not sure if there might be any clues in there?

Reply:
Good shout @DMRobertson, this seemed to work to get both versions side by side:
```
$ poetry add --extras tls git+https://github.com/twisted/twisted.git#trunk

Updating dependencies
Resolving dependencies... (24.3s)

Writing lock file

Package operations: 0 installs, 1 update, 0 removals

  • Updating treq (22.2.0 -> 15.1.0)

$ poetry add treq@22.2.0

Updating dependencies
Resolving dependencies... (12.4s)

Writing lock file

Package operations: 0 installs, 1 update, 0 removals

  • Updating treq (15.1.0 -> 22.2.0)
```

---

But still getting `[BUG]  twisted (rev trunk) is not satisfied.` in the CI when linting?!? https://github.com/matrix-org/synapse/runs/7666653208?check_suite_focus=true

Not sure what that means even. `lint.sh` passes locally 🤔